NeMo getting started guide

smythp opened the pull request:
## Type of change

Adds a getting started guide for the NeMo Chainguard Image

## Acceptance Criteria / Blockers

Currently this image has CVEs that need to be remediated. I'll keep this as a draft until we can complete this with a 0 or near-0 CVE count. It can still be reviewed, however.

## Testing

The current example can run on a CPU without much problems, but a deep learning environment should also be used to test. I have instructions for setting one up [here](https://github.com/chainguard-images/images/blob/main/images/pytorch-cuda12/TESTING.md#creating-a-deep-learning-vm).

## Notes

I'd be open to adding a finetuning example, but we're waiting on getting some deps into the image. We can wait for more stability / customer requests.

I added a deep learning blurb, I'll put it in the pytorch example after this merges.

I added a widget to play audio files to our shortcodes for this.

I added a details widget that is open by default that I can use for scanner output.

References https://github.com/chainguard-dev/internal/issues/3731

smythp committed:
NeMo draft, audio shortcode, deep learning blurb, output WAV

Signed-off-by: Patrick Smyth <patrick.smyth@chainguard.dev>

smythp committed:
New sections, comparison

Signed-off-by: Patrick Smyth <patrick.smyth@chainguard.dev>

smythp committed:
Added open details shortcode, formatted comparison

Signed-off-by: Patrick Smyth <patrick.smyth@chainguard.dev>

smythp committed:
Updated resources, copy edit

Signed-off-by: Patrick Smyth <patrick.smyth@chainguard.dev>

smythp committed:
Merge branch 'main' into nemo-getting-started

erikaheidi commented on content/chainguard/chainguard-images/getting-started/nemo/index.md:
````diff
@@ -0,0 +1,3013 @@
+After imports are complete, you should see a large amount of output as NeMo pulls models and works through the steps in the script (tokenizing, generating a spectrogram, generating audio, and writing audio to disk). On completion, the script outputs a `test.wav` file. Because we mounted a volume, this file should now be present in the working directory of your host machine.
+
+```bash
+$ ls
````
This can look a bit confusing with the command and output together in the same block. I don't think you need to add the `ls` command tbh. If you want to, it's best to have them in separate blocks to optimized rendering:

```shell
ls
```

```
output
```

smythp replied:
I removed one `ls` command and left the other, as I feel like it shows the final state more clearly than another paragraph, etc. I hear you on the $, I removed them from the docker commands as they do make copying a pain. We might want to talk as a team about options for including prompts in a non-copyable way on CG Academy, something like a smart shortcode or similar.

erikaheidi commented on content/chainguard/chainguard-images/getting-started/nemo/index.md:
````diff
@@ -0,0 +1,3013 @@
+test.wav  tts.py
+```
+
+Note that we ran the above script as root. This allows us to share the script and output `.wav` file between the host and container. Remember not to run your image as root in a production environment.
````
I think this should go right below the command

erikaheidi commented on content/chainguard/chainguard-images/getting-started/nemo/index.md:
```diff
@@ -0,0 +1,3013 @@
+
+{{< audio src="tts-example.wav" caption="Output from the TTS script" >}}
+
+## Advantages of Chainguard Images for Production ML
```
I'm not quite sure the getting started guide is the best place for the image comparison, because this data fluctuates a lot and we won't be able to keep it up-to-date. I would advise to remove this or simplify removing the itemized CVEs part. this data should be in the [vul comparison](https://edu.chainguard.dev/chainguard/chainguard-images/vuln-comparison/) section IMO and then you could link to that from here.

smythp replied:
A version of this is going out to beta clients and they want this in for now. I'll look into (re)moving this when things settle down.

erikaheidi commented on content/chainguard/chainguard-images/getting-started/nemo/index.md:
I would merge this and the next sections under a `## Final Considerations` header where you give advice on next steps for exploration and resources to learn more.

smythp replied:
I took out the script review section and added the content to other sections or removed it. :+1:

erikaheidi reviewed (changes requested):
Hey Patrick! I had some issues to run the example using my GPU. I am not sure what I did wrong! In any case, I did the editorial review and have some suggestions to improve the flow of the article and make it easier to follow. I will try to debug my issue with Docker for the tech run.

smythp commented on content/chainguard/chainguard-images/getting-started/nemo/index.md:
````diff
@@ -0,0 +1,3013 @@
+  cgr.dev/chainguard/nemo:latest
+```
+
+These options allow access to connected GPU, allocate more shared memory to the container, and put an upper bound on container memory use. 
````
Changed to make the language more specific as in your example, but kep the parallel structure. :+1:

smythp commented on content/chainguard/chainguard-images/getting-started/nemo/index.md:
```diff
@@ -0,0 +1,3013 @@
+
+These options allow access to connected GPU, allocate more shared memory to the container, and put an upper bound on container memory use. 
+
+Running the above for the first time may take a few minutes to pull the NeMo Chainguard Image, currently 4.05 GB. Once the image runs, you will be interacting with a bash shell in the running container. Enter the following commands at the prompt to check the availability of your GPU.
```
I agree that it would be good to include this, though getting CUDA to work on a local machine is (very) nontrivial. Mac OS is not supported at all if that's the OS you're using. I've written a new blurb that will be included in a prerequisites section. It links to guides on setting  up CUDA for Linux, Windows, GCP, and AWS. As a practical matter I recommend testing on GCP, it's pretty uphill getting this working without preselecting your hardware, OS, and so on for this task. I have updated [this page](https://github.com/chainguard-images/images/blob/main/images/pytorch-cuda12/TESTING.md) with instructions for setting up GCP.

If you were using Mac OS, it looks like that isn't supported at all. :(

Here's the blurb:

<p>
  <a href="https://developer.nvidia.com/cuda-toolkit">Compute Unified Device Architecture (CUDA)</a> is a parallel computing platform developed by NVIDIA. To take advantage of connected NVIDIA GPUs, you’ll need to follow the setup instructions for your local machine or create a CUDA-enabled instance on a cloud provider.
</p>
<p>
  To set up CUDA on your local machine, follow the installation instructions for <a href="https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html">Linux</a> or <a href="https://docs.nvidia.com/cuda/cuda-installation-guide-microsoft-windows/index.html">Windows</a>. CUDA is not currently supported on Mac OS.
</p>
<p>
  Google Cloud Platform provides CUDA-ready <a href="https://cloud.google.com/deep-learning-vm">deep learning instances</a>, including <a href="https://cloud.google.com/deep-learning-vm/docs/pytorch_start_instance">PyTorch-specific instructions</a>. Amazon Web Services also provides CUDA-ready <a href="https://aws.amazon.com/machine-learning/amis/">deep learning instances</a>.
</p>

erikaheidi replied:
Thank you for adding this section! As someone totally newbie to this area, I did not know things wouldn't work out of the box, so I think it's better to make sure users are aware of everything needed to test the image. I am on Linux, going now through the installation instructions to try to run it locally.

erikaheidi replied:
I was able to run it on GCP 🎉

smythp commented on content/chainguard/chainguard-images/getting-started/nemo/index.md:
```diff
@@ -0,0 +1,3013 @@
+
+This guide is primarily designed for use in an environment with access to one or more NVIDIA GPUs. However, NVIDIA NeMo is built on [PyTorch Lightning](https://lightning.ai/docs/pytorch/stable/), which supports a wide variety of [accelerators](https://pytorch-lightning.readthedocs.io/en/1.1.8/accelerators.html), or interfaces to categories of processing units (CPU, GPU, TPU) or high-level cluster ...
+
+## Testing Access to GPUs
```
I don't think this tutorial lends itself to steps in the headings. However, given that, the language (first step) in that first paragraph is confusing. I've edited it to remove that language, thanks!

smythp commented on content/chainguard/chainguard-images/getting-started/nemo/index.md:
````diff
@@ -0,0 +1,3013 @@
+You should now be in a working directory with `tts.py`.
+
+```bash
+$ ls
````
I've removed the `ls` section and added in a little more language in the prose to get across the same thing.

smythp committed:
Edits based on Erika's review

Removed script overview section, moved content to other sections

Added CUDA blurb (can also add it to pytorch getting started) and
placed it in new prereqs section

Removed most $ in bash commands

Signed-off-by: Patrick Smyth <patrick.smyth@chainguard.dev>

erikaheidi reviewed (approved):
I created a VM on GCloud and tested everything, works like a charm, and it's super cool. I wasn't able to run it locally because of my Ubuntu version (23.10, tried to install the kit for 20.4 but it didn't work).

smythp commented:
Awesome, thanks Erika, and thanks for the editorial review!

@ltagliaferri, what do you think of taking out  the image comparison section for now and publishing? We could also wait a couple days for CVE mitigation, there's an active effort.

smythp committed:
Removed comparison section

Signed-off-by: Patrick Smyth <patrick.smyth@chainguard.dev>
